pkg/apps/packagekit.js

```diff
@@ -0,0 +1,185 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+var cockpit = require("cockpit");
+var $ = require("jquery");
+
+var client = cockpit.dbus("org.freedesktop.PackageKit", { superuser: "try" });
+
+var PK_STATUS_ENUM_WAIT = 1;
+var PK_STATUS_ENUM_WAITING_FOR_LOCK = 30;
+
+var PK_FILTER_INSTALLED   = (1 << 2);
+var PK_FILTER_NEWEST      = (1 << 16);
+var PK_FILTER_ARCH        = (1 << 18);
+var PK_FILTER_NOT_SOURCE  = (1 << 21);
+
+function transaction(method, args, progress_cb, package_cb) {
+    var defer = cockpit.defer();
+
+    client.call("/org/freedesktop/PackageKit", "org.freedesktop.PackageKit", "CreateTransaction", [ ]).
+        done(function(path_result) {
+            var tr = client.proxy("org.freedesktop.PackageKit.Transaction", path_result[0]);
```

Comment:
Does that proxy and its event listeners ever get cleaned up? In updates I ran into some trouble after doing more than two or three transactions, the multiple proxies confused and locked up PK (which is why I rewrote it to not use proxies at all).

Reply:
There is no way to clean up a proxy.  It will be GCed eventually, but I think that is not what you mean, right?

Reply:
Correct. Don't get me wrong, if/as long as the proxies work for you, it's okay. I just wanted to point out that I had troubles with too many signal listeners (and perhaps introspection calls), it caused PK lockups which broke updates and required a daemon restart.

Reply:
I see... hmm.  What about leaving things as they are and when the tests show the same PackageKit crashes as they did for the OS Updates, I'll do something about it?

Reply:
Sure, sounds good.